### PR TITLE
WIP: mc_pos_control: triplet handling for mc

### PR DESCRIPTION
**Pull request by Stifael**
currently the handling of the triplet setpoint in mc_pos_control_main is rather confusing. In addition, the coupling of vtol to mc makes it also hard to distinguish what is important for vtol and what not. Following the discussion from here #6681, the triplet handling for simple point to point motions are implemented too complicated. Setting the desired current setpoint as the desired position is enough. There should not be any additional slewrate applied to the position. It is completely fine to have a position setpoint step. The velocity limit is applied  already at the output of the position controller. 
For the handling of waypoints, I can see the benefit of incorporating the previous and next setpoint, but I would completely split the implementation depending on vtol and normal mc (maybe that is already the case, but it is not obvious to me). Also, instead of having this position setpoint adjustment, I suggest to apply normal position control perpendicular to the line between previous and current setpoint and velocity feedforward in the velocity controller. the velocity feedforward vector can be computed from bezier curve which also incorporates max acceleration. I know how to do trajectory based on bezier curve for mc, but not for vtol and that is another reason why I would split the implementation entirely. obviously with vtol I mean when in fixedwing mode.

### CODE REVIEW

File: src/modules/mc_pos_control/mc_pos_control_main.cpp [hunk 1]
```diff
@@ -1429,9 +1429,11 @@ void MulticopterPositionControl::control_auto(float dt)
 
 	bool current_setpoint_valid = false;
 	bool previous_setpoint_valid = false;
+	bool next_setpoint_valid = false;
```
Comment (dagar) [inline src/modules/mc_pos_control/mc_pos_control_main.cpp]: Can we drop these and use the valid bool each position setpoint already has?

**Comment by dagar** [comment]
I don't think I follow what you're saying about VTOL.

Currently VTOL is running both MC (pos + att) and FW (pos + att) at all times and the VTOL controller is responsible for engaging/disengaging MC or FW actuator outputs and handling the transition between VTOL modes. In MC mode a VTOL is a regular multicopter, and likewise for FW.

MC pos is only aware of the existence of VTOL so that things are properly reset right before transitioning from FW to MC.

**Comment by Stifael** [comment]
I guess what confused me is this line https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/mc_pos_control_main.cpp#L1417

**Comment by dagar** [comment]
Centralize any VTOL related init in the same spot? https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/mc_pos_control_main.cpp#L2330

**Comment by matanhavi** [comment]
I have tested the next scenario, changing the simulator altitude to negative value and seen critical bad behaviors.
I have done that by changing the line:
[static const double alt_zurich =](https://github.com/PX4/sitl_gazebo/blob/2c64ddd7419a96df36809411959f87d7b88e6e06/src/gazebo_mavlink_interface.cpp#L35)

On some cases, the drone start to change altitude rapidly after its seems stabilized 
On some cases its takeoff, and doesn't climb, just stay close to ground (for example, for -50m)
-different altitude seem to behave different, 200 seems to be ok for some reason

**Comment by matanhavi** [comment]
I also noticed that on landing command POSITION_TARGET_LOCAL_NED.z is set to global altitude of home position. On the previous behavior it wasn't.
What If landing altitude is not the takeoff alt?

**Comment by Stifael** [comment]
@matanhavi: thanks for your testing. indeed, it needs more improvements than just the slewrate in positiont setpoint. by the way, if you have time you are more than welcomed to fix whatever you see is wrong during your testing.
about the second issue:
> ...POSITION_TARGET_LOCAL_NED.z is set to global altitude of home position
currently the altitude for landing is only used to decrease the velocity in z direction, but not to detect the landing. it is true that the vehicle will also fly slower in the z direction if you do a takeoff in the mountains and then fly downwards toward the valley. This problem could be resolved if the vehicle would have a reliable source of height detection relative to ground.

**Comment by Stifael** [comment]
@dagar: 
Can you explain to me the meaning of rotary wing? what is this flag doing in mc pos control? is that only needed for transition?

**Comment by dagar** [comment]
Rotary wing = MC

Always true for a regular multicopter, always false for a regular plane.
For VTOL it's true in regular hover, false in fixed wing, and true during
transitions.

Outside of VTOL you might have it to stop the MC position controller from
running/publishing if someone started it on a plane.

**Comment by matanhavi** [comment]
@Stifael can you better explain what you expect the bezier curve to do? 
Is it improvement of current position logic? 
or a "fix", something is "wrong" in current logic?
or it might be "option"? if it is, it should not fully replace current logic.

about POSITION_TARGET_LOCAL_NED.z is set to global altitude of home position, I know that the z sp isn't the position where the motor will stop. I did mention that the value is a bit different that previous behavior, and it need to be noticed.

I also recommend you to test any logic related to position sp around the equator, Greenwich, and +-180deg.

and because its mathematical logic, check MCU %, make sure they doesn't rise too much. @dagar, are there any cpu/ram test between versions?

**Comment by Stifael** [comment]
It will replace the auto logic, but im working on a simpler version to at least fix some of the obvious bugs.
The purpose of the bezier is to replace position setpoint slewrate, which has some fundamental problems due to its feedforward implementation. There is also no reason why you need to generate setpoints on 2d. The bezier curve is one method to compute trajectory from which position, velocity and acceleration can be derived from depending on current state. However, this logic should in fact be implemented a layer up in the navigator, which requires even more changes. Thats why I started to put it in the mc for now.

**Comment by Stifael** [comment]
@matanhavi: about altitude
If battery is low, the vehicle cannot exceed a certain limit. What was the battery status in sitl?

**Pull request closed without merging**